src/library_int53.js

```diff
@@ -93,23 +93,27 @@ mergeInto(LibraryManager.library, {
     return HEAPU32[ptr>>2] + HEAPU32[ptr+4>>2] * 4294967296;
   },
 
-  // Converts the given signed 32-bit low-high pair to a JavaScript Number that can
-  // represent 53 bits of precision.
+  // Converts the given 32-bit low-high pair to a signed 53-bit integer with
+  // the sign of the high word (and represented as a JavaScript Number).
+  // The signedness of the low word doesn't matter; it will be bit-cast to u32.
   // TODO: Add $convertI32PairToI53Signaling() variant.
   $convertI32PairToI53: function(lo, hi) {
 #if ASSERTIONS
-    // This function should not be getting called with too large unsigned numbers
-    // in high part (if hi >= 0x7FFFFFFFF, one should have been calling
-    // convertU32PairToI53())
-    assert(hi === (hi|0));
+    // Value should not be outside the 53-bit integer range.
+    assert(hi >= -0x200000 && hi <= 0x200000);
```

Comment:
This was not the original intent of the `convertI32PairToI53()` function.

The intended semantics of `convertI32PairToI53()` is to behave identical to `writeI53ToI64()`, `readI53FromI64()` and `readI53FromU64()` in that numbers in the categories 1, 2 and 3 in [here](https://github.com/emscripten-core/emscripten/blob/e274e49bfb2022313f5697dc5560c19f5f13d81a/tests/core/test_int53.c#L176) are exactly convertable, and numbers in the categories 4 and 5 there are also convertable, but with a round that is incurred.

I think this change would prevent this function from being called on numbers from categories 2,3,4 and 5? It looks like the core test suite run agrees.

I agree with your changed comment that the naming `I32Pair` and `U32Pair` in these functions is perhaps not the best. The idea here is to take a pair of 32-bit signed or unsigned lo-hi numbers, then interpret those bits as a signed 64-bit integer, returned as a Number. The assertion here is to enforce a convention that users would not read such signed 64-bit integers off the heap using `HEAPU32`, which would result in the `hi` value being large unsigned (hinting towards mixed interpretation in caller's end) - but instead the caller is expected to interpret the i32 pair as signed already (maybe calling `hi=hi|0` themselves if they source the value from somewhere odd place)

Do you have a specific number usage in mind here that prompted you to do this change? Maybe we could look at a concrete example to see if there's a brainfart in the design somewhere.

Reply:
edit: I think the correct check would be `(result >>> 0) == (lo >>> 0)`.

Just pushed a draft change that adds that assertion.